docs/adr/009_lns_affinity.md

```diff
@@ -0,0 +1,262 @@
+   memory that it was the losing gateway for this DevEui**.
+1. When message B gets send (with a higher frame counter), assuming that this time LNS2 gets it
+   first it **checks again its local dictionary it's not the preferred LNS for this device and
+   therefore delays itself X ms before contacting the FunctionBundler**.
+   - Here we can *not* simply drop as LNS1 might not be available anymore (due to a crash, device not
+     in range etc).
+1. This delay gives LNS1 a time advantage to reach the FunctionBundler first and win the race again, failing
+    back to the previous case of message A. The active connection stays with LNS1.
+1. If this delay is not sufficient for LNS1 to win the race, LNS2 will contact the FunctionBundler which
+   now awards LNS2 as the "winning" LNS. LNS2 will process message upstream (therefore the active
+   connection will switch to it) and **removes the "losing flag" from the in-memory store**.
+1. **The Function also proactively informs LNS1** that it's not anymore the winning LNS for this
+   device. The reason why we do this proactively is that otherwise its Edge Hub will try to
+   reconnect to IoT Hub even if there is no more messages picked up from LNS1 (out of range roaming
+   client). Additionally, **LNS1 would not need to refresh its LoRaDeviceCache anymore for this
+   device** (see [previous section about caching](#data-flow-loradevice-not-in-loradevicecache)).
+1. Direct method and should be retried (e.g. in memory retries or durable Function): direct
+   method could fail due to the module being or Edge Hub being down, other connectivity issue
+   etc. Is in memory retries sufficient here? ❔
+1. If LNS1 in the meantime gets message B and contacts the FunctionBundler, it will let it know
+   that it lost the race for this frame counter and must therefore drop the message, **mark
+   itself as the losing LNS and close the connection**.
+
+NB: the FunctionBundler is not called in certain topologies e.g. when multiple LBSs are connected to
+the same LNS but these topologies are not relevant for the issue here).
+
+#### Handling Class C devices downstream direct messages
+
+When we send a downstream message using Direct method via the Function, the Function is responsible
+for choosing the prererred gateway based on its existing knowledge of the preferred gateway. Due to
+that we can be sure that we have the active connection and can therefore send downstream messages.
+
+#### Implementation
+
+The LNS in-memory dictionary is mapping DevEuis to a flag indicating whether we are the winning or losing LNS.
+
+- Where should we store locally this structure ❔
+
+  - LNS LoRaDeviceClient.ConnectionManager since all of the operations pass through it.
+  - LoRaDevice: in the case of Join requests we don't yet have a LoRaDevice and would need to create
+    one without fetching the twin
+  - LoRaDeviceClient
+
+#### Delay on the LNS itself or on the Function ❔
+
```

Comment:
I think an advantage of doing it locally, might be the possibility to compute the artifical delay programmatically. ie. If you lost a lot of time in the first phase of the message you could reduce the delay to avoid that the new gateway miss the receive window.

Reply:
Listed it as a potential advantage.